Release TLS1.3 Session Resumption (#2877)
aws/s2n-tls: docs/USAGE-GUIDE.md

```c
typedef enum { S2N_CERT_AUTH_NONE, S2N_CERT_AUTH_REQUIRED, S2N_CERT_AUTH_OPTIONAL } s2n_cert_auth_type;
```

> **s2n_cert_auth_type** is used to declare what type of client certificate authentication to use.

Reviewer: You clearly got your spellchecker working again!

Reviewer 2: is used - change to used

Author: @Lucy-writer, "is used" is in a lot of places in this doc, we should probably remove them all at once rather than in this pr, if we do want this changed.

s2n-tls includes support for resuming from cached SSL/TLS session, provided 
the caller sets (and implements) three callback functions.

### s2n\_config\_set\_session\_cache\_onoff

> **s2n_config_set_session_cache_onoff** enables and disables the session cache. In order to use the session cache feature you will need to set a ticket key with **s2n_config_add_ticket_crypto_key** to encrypt session state inside the cache.

Reviewer 3: From the description it's a bit not clear what a session cache feature is? Is it a cache when we're in `S2N_CLIENT` mode, so that s2n implicitly caches and resumes sessions?

Reviewer 2: remove "that will be" - all instances
The first pointer will be set to - change to "The first pointer sets "
In order to use the session cache feature you will need - remove "will"

Author: I'm going to remove this addition. The session cache doesn't have anything to do with this release, and there's probably more documentation that needs to go into this feature anyways. I'll create an issue for us to deal with this later.

> **s2n_config_set_session_cache_onoff** enables and disables the session cache. In order to use the session cache feature you will need to set a ticket key with **s2n_config_add_ticket_crypto_key** to encrypt session state inside the cache.
> 

Reviewer: I think we avoid "you". So
```suggestion
**s2n_config_set_session_cache_onoff** enables and disables the session cache. In order to use the session cache feature, a user must set a ticket key with **s2n_config_add_ticket_crypto_key** to encrypt session state inside the cache.

```

s2n-tls includes support for resuming from cached SSL/TLS session, provided 
the caller sets (and implements) three callback functions.

### s2n\_config\_set\_session\_cache\_onoff

**s2n_config_set_session_cache_onoff** enables the session cache. In order to use the session cache feature

> you will need to set a ticket key with **s2n_config_add_ticket_crypto_key** to encrypt session

Reviewer 4: what will this function return if a ticket_crypto_key has not been set?

Author: Heh. It would return successful. It's just a function to indicate a user wants to use session cache, it doesn't check if a ticket key has been added.

Reviewer 4: that seems weird. so a user could call this function and since it returns successfully they could go on to assume caching works. should this functions return an error if the key is not set? that would prevent an invalid state?

was there a reason for not returning an error?

Author: I think the general idea of the s2n on/off functions is to indicate which type of resumption the user wants to do, rather than checking that the user has the correct configurations to do that specific type of resumption. 🤷‍♀️ They will throw a S2N_ERR_NO_TICKET_ENCRYPT_DECRYPT_KEY error if you try to resume without adding a ticket key.

**s2n_connection_set_session** de-serializes the session state and updates the connection accordingly.

**s2n_connection_get_session** serializes the session state from connection and copies into the **session** buffer and returns the number of bytes that were copied. The output of this function depends on whether session ids or session tickets are being used for resumption.

> If the first byte in **session** is 1, then the next 2 bytes will contain the session ticket length, followed by session ticket and session state. In versions TLS1.3 and greater, (which allows multiple session tickets), the most recent session ticket received will be used. Note that the size of the tickets increased between TLS1.2 and TLS1.3.

Reviewer: You might want to just say that the size of the tickets varies, since not all TLS1.3 tickets are the same size either.

**s2n_connection_set_session** de-serializes the session state and updates the connection accordingly.

> **s2n_connection_get_session** serializes the session state from connection and copies into the **session** buffer and returns the number of bytes that were copied. The output of this function depends on whether session ids or session tickets are being used for resumption.

Author: I added a little blurb about how the TLS1.3 tickets are larger than TLS1.2 ticket. I don't think the user needs to know the differences in format. They shouldn't be messing with those values anyway.

**s2n_connection_set_session** de-serializes the session state and updates the connection accordingly.

> **s2n_connection_get_session** serializes the session state from connection and copies into the **session** buffer and returns the number of bytes that were copied. If the first byte in **session** is 1, then the next 2 bytes will contain the session ticket length, followed by session ticket and session state. If the first byte in **session** is 0, then the next byte will contain session id length, followed by session id and session state. This function was revamped in TLS1.3 to retrieve the last session ticket received by the client.

Reviewer: Maybe not revamped? That feels like it might go out of date. Let's just describe the behavior.
```suggestion
**s2n_connection_get_session** serializes the session state from connection and copies into the **session** buffer and returns the number of bytes that were copied. If the first byte in **session** is 1, then the next 2 bytes will contain the session ticket length, followed by session ticket and session state. In TLS1.3 (which allows multiple session tickets), the most recent session ticket received will be used. If the first byte in **session** is 0, then the next byte will contain session id length, followed by session id and session state.
```

Reviewer 4: maybe break this into multiple paragraphs and some more info on return type:

```
...number of bytes that were copied. This function can return either a session_id or session_ticket (what determines one over the other?). The first byte in session is used to indicate the return type.

If the first byte i...
```

Author: I tried to break it up a little and clarify what is going on.

**s2n_connection_is_session_resumed** returns 1 if the handshake was abbreviated, otherwise returns 0, for tls versions < TLS1.3.

## Additional TLS1.3 Session Resumption Related calls

> Session resumption works differently in versions TLS1.3 and higher. Session ticket messages are now sent after the
> handshake in "post-handshake" messages. While some of the TLS1.2 session resumption APIs are still relevant for TLS1.3

Reviewer: ```suggestion
Session resumption works differently in versions TLS1.3 and higher. Session ticket messages are now sent after the
handshake in "post-handshake" messages, and multiple session tickets may be issued for the same connection. While some of the TLS1.2 session resumption APIs are still relevant for TLS1.3
```

Reviewer 4: is there a good way to tell which apis are still relevant for tls1.3 and which are not? is it documented somewhere?

session resumption, additional APIs are needed to utilize all the capabilities of TLS1.3 session resumption.

```c
int s2n_config_set_initial_ticket_count(struct s2n_config *config, uint8_t num);

> 

Reviewer: Why a newline?

int s2n_connection_add_new_tickets_to_send(struct s2n_connection *conn, uint8_t num);
int s2n_connection_set_server_keying_material_lifetime(struct s2n_connection *conn, uint32_t lifetime_in_secs);

typedef int (*s2n_session_ticket_fn)(struct s2n_connection *conn, void *ctx, struct s2n_session_ticket *ticket);
int s2n_config_set_session_ticket_cb(struct s2n_config *config, s2n_session_ticket_fn callback, void *ctx);
int s2n_session_ticket_get_data_len(struct s2n_session_ticket *ticket, size_t *data_len);
int s2n_session_ticket_get_data(struct s2n_session_ticket *ticket, size_t max_data_len, uint8_t *data);
int s2n_session_ticket_get_lifetime(struct s2n_session_ticket *ticket, uint32_t *session_lifetime);
```

**s2n_config_set_initial_ticket_count** sets the initial number of session tickets to send.

**s2n_connection_add_new_tickets_to_send** increases the number of session tickets to send by **num**.

**s2n_connection_set_server_keying_material_lifetime** sets the keying material lifetime for session tickets so that one

> session doesn't get re-used ad infinitum.

Reviewer: "one session doesn't get re-used" might not be descriptive enough to explain why you'd set this. We might want to be more specific: one authentication? One full handshake? We don't want a session to outlive its original certificate?

Author: Added more descriptive blurb

**s2n_connection_is_session_resumed** returns 1 if the handshake was abbreviated, otherwise returns 0, for tls versions < TLS1.3.

## Additional TLS1.3 Session Resumption Related Calls

> Session resumption works differently in versions TLS1.3 and higher. Session ticket messages are now sent after the handshake in "post-handshake" messages, and multiple session tickets may be issued for the same connection. While some of the TLS1.2 session resumption APIs are still relevant for TLS1.3 session resumption, additional APIs are needed to utilize all the capabilities of TLS1.3 session resumption.

Reviewer 3: Do we send session tickets immediately after handshake only? Or s2n may send them in the middle of the session too?

Reviewer 2: returns the number of bytes that were copied. - remove "that were"
While some of the TLS1.2 session resumption APIs are still relevant for TLS1.3 session resumption, additional APIs are needed to utilize all the capabilities of TLS1.3 session resumption. - Change to "APIs have revelance" and "you need additional APIs to utilize"

Author: I added a few more clarifying comments about when session tickets are sent.

```c
int s2n_config_set_initial_ticket_count(struct s2n_config *config, uint8_t num);
int s2n_connection_add_new_tickets_to_send(struct s2n_connection *conn, uint8_t num);
int s2n_connection_set_server_keying_material_lifetime(struct s2n_connection *conn, uint32_t lifetime_in_secs);

typedef int (*s2n_session_ticket_fn)(struct s2n_connection *conn, void *ctx, struct s2n_session_ticket *ticket);
int s2n_config_set_session_ticket_cb(struct s2n_config *config, s2n_session_ticket_fn callback, void *ctx);
int s2n_session_ticket_get_data_len(struct s2n_session_ticket *ticket, size_t *data_len);
int s2n_session_ticket_get_data(struct s2n_session_ticket *ticket, size_t max_data_len, uint8_t *data);
int s2n_session_ticket_get_lifetime(struct s2n_session_ticket *ticket, uint32_t *session_lifetime);

> ```

Reviewer 5: Could we follow the same format that we do for the other API descriptions in the Usage Guide. It seems like Session Resumption APIs are the outlier. For API descriptions the rest of the Usage Guide follows: 

- API Title 
- API Snippet 
- Description 

For ex: https://github.com/aws/s2n-tls/blob/main/docs/USAGE-GUIDE.md#s2n_config_new

```
### s2n\_config\_new

```c
struct s2n_config * s2n_config_new();
 `` `
 
s2n_config_new** returns a new configuration object suitable for associating certs and keys.
This object can (and should) be associated with many connection objects. 
```

Reviewer: Idk, we're not consistent, especially for chunks of closely related APIs. For example, https://github.com/aws/s2n-tls/blob/main/docs/USAGE-GUIDE.md#s2n_mem_set_callbacks, https://github.com/aws/s2n-tls/blob/main/docs/USAGE-GUIDE.md#s2n_connection_set_send_cb, https://github.com/aws/s2n-tls/blob/main/docs/USAGE-GUIDE.md#asynchronous-private-key-operations-related-calls.

Let's do what works / is readable.

Author: Yeah, I'm trying to follow the format of the TLS1.2 session resumption docs so they fit together cohesively. I would prefer if we just stuck to one style though.

> **s2n_config_set_initial_ticket_count** sets the initial number of session tickets to send.

Reviewer 3: If not called, what is the default value? 1? What is the good value the user should configure this too? What are the drawbacks of setting it too low or too high?

Author: I added what the default value is. I'm not sure specifying "you can send up to 2^16 tickets" is that helpful for users. You should send as many tickets as you want the client to have 🤷‍♀️? The only cost is client storage space.

**s2n_connection_add_new_tickets_to_send** increases the number of session tickets to send by **num**.

> **s2n_connection_set_server_keying_material_lifetime** sets the keying material lifetime for session tickets. Use this to ensure session tickets don't get reissued past the lifetime of the certificate used to authenticate the original full handshake.

Reviewer 3: Same here - what is the default value? And if we're only worried about the lifetime of the certificate, can't s2n use NotAfter date on certificate instead?

Author: I added the default current default value. I think Lindsay might be changing it before release so that it is the server's certificate lifetime, as that is a good idea.

> **s2n_session_ticket_fn** is invoked whenever a client receives a session ticket. Use this callback in conjunction with the **s2n_session_ticket** getters to get the serialized ticket data and related information. A **ctx** pointer is provided to let a user pass state to the callback, if needed.

Reviewer: I think "user" is probably clearer here. Combining "caller" with "callback" doesn't seem quite right, considering S2N is actually the caller of a callback.

> **s2n_config_set_session_ticket_cb** sets the session ticket callback function to be invoked whenever the client receives

Reviewer 3: Can the malicious server just send tickets all the time potentially causing issues to the client if callback implementation is expensive or keeps allocating memory? If so should we note about the risks in documentaiton?

Author: Added a note about the risks. I think this is relatively low-risk, given that all of this is happening post-handshake, but clients should be aware the server can send a lot of tickets.

a session ticket from the server.

**s2n_session_ticket_get_data_len** takes a s2n_session_ticket object and retrieves the number of bytes needed to store the session ticket. Use this to allocate enough memory for the session ticket in **s2n_session_ticket_get_data**.

**s2n_session_ticket_get_data** takes a s2n_session_ticket object and copies the serialized session ticket data into the

> **data** buffer. For this reason **max_data_len** should be set to the maximum amount of bytes that can be copied into

Author: The function will error if you don't have enough memory to store the ticket. I did change the `should` phrasing though.

the **data** buffer.

> **s2n_session_ticket_get_lifetime** takes a s2n_session_ticket object and retrieves the lifetime of the ticket in seconds.

Author: s2n_session_ticket_get_lifetime just gets the lifetime of the ticket, it wouldn't change whether or not the ticket was expired. I also don't think this is a problem, the s2n_session_ticket object doesn't live past the callback, so it would be really unlikely to try to get the lifetime of a ticket already expired.
